### PR TITLE
Add tests on prehash and encode_bin

### PR DESCRIPTION
**Pull request by icristescu**

### CODE REVIEW

File: test/irmin-tezos/test_hashes.ml [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+let check_pre_hash pre_hash v checks =
+  let counter = ref 0 in
+  pre_hash v (fun x ->
+      match List.nth_opt checks !counter with
+      | None -> Alcotest.fail "No more calls to pre_hash left"
+      | Some (msg, expected) ->
+          Fmt.epr "counter at %d\n%!" !counter;
+          check_prehash msg expected x;
+          incr counter)
```
Comment (Ngoguey42) [inline test/irmin-tezos/test_hashes.ml:27]: Does a `!counter = List.length checks` test makes sense here?

File: test/irmin-tezos/test_hashes.ml [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+      [
+        ("hash of root node", hash_root_small_tree);
+        ("len of parents", "01");
+        ( "parent hash",
+          "bf50614cb5f7c3a35e613ed5abd6ea1e2619a0e832e88bbbe436b74a6151ed03" );
+        ("date", "0000000000000000");
+        ("len of author", "05");
+        ("author", "54657a6f73");
+        ("len of message", "00");
+        ("message", "");
```
Comment (Ngoguey42) [inline test/irmin-tezos/test_hashes.ml:171]: Great trick!

File: test/irmin-tezos/test_hashes.ml [hunk 3]
```diff
@@ -0,0 +1,258 @@
+open! Import
+
+let root = Filename.concat "_build" "test-irmin-tezos"
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply test/irmin-tezos/test_hashes.ml:3]: (We should probably put that bit of code inside `irmin-test`.)

File: test/irmin-tezos/test_hashes.ml [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        Alcotest.test_case "contents hash" `Quick (fun () ->
+            Lwt_main.run (Test_tezos_conf.contents_hash ()));
+        Alcotest.test_case "inode_values hash" `Quick (fun () ->
+            Lwt_main.run (Test_tezos_conf.inode_values_hash ()));
+        Alcotest.test_case "inode_tree hash" `Quick (fun () ->
+            Lwt_main.run (Test_small_conf.inode_tree_hash ()));
+        Alcotest.test_case "commit hash" `Quick (fun () ->
+            Lwt_main.run (Test_tezos_conf.commit_hash ()));
+      ] );
+  ]
```
Comment (CraigFe) [inline test/irmin-tezos/test_hashes.ml:289]:
```suggestion
  let tc name f = Alcotest.test_case name `Quick (fun () -> Lwt_main.run (f ())) in
  [
    ( "hashes",
      [
        tc "contents hash" Test_tezos_conf.contents_hash;
        tc "inode_values hash" Test_tezos_conf.inode_values_hash;
        tc "inode_tree hash" Test_small_conf.inode_tree_hash;
        tc "commit hash" Test_tezos_conf.commit_hash;
      ] );
  ]
```

File: test/irmin-tezos/test_hashes.ml [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+  val entries : int
+  val stable_hash : int
+end) =
+struct
+  module Store = struct
+    module V1 = struct
+      let version = `V1
+    end
+
+    module Maker = Irmin_pack.Maker_ext (V1) (Conf)
```
Comment (CraigFe) [inline test/irmin-tezos/test_hashes.ml:49]:
```suggestion
    module Maker = Irmin_pack.Maker_ext (Irmin_pack.Version.V1) (Conf)
```

File: test/irmin-tezos/test_hashes.ml [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+let hash_zero =
+  "d81e60258ecc8bd7064c8703888aececfc54e29ff94f7a2d9a84667a500548e1"
+
+let bindings steps = List.map (fun x -> ([ x ], zero)) steps
+
+module Test (Conf : sig
+  val entries : int
+  val stable_hash : int
+end) =
```
Comment (CraigFe) [inline test/irmin-tezos/test_hashes.ml:42]:
```suggestion
module Test (Conf : Irmin_pack.Conf.S) =
```

File: test/irmin-tezos/test_hashes.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (icristescu) [reply test/irmin-tezos/test_hashes.ml]: moved the tests to irmin-pack

File: test/irmin-tezos/test_hashes.ml (hunk 1 above)
Comment (icristescu) [reply test/irmin-tezos/test_hashes.ml]: yes! thanks

File: test/irmin-pack/test_pack.ml [hunk 7]
```diff
@@ -682,4 +682,5 @@ let misc =
     ("layers", Layered.tests);
     ("inodes", Test_inode.tests);
     ("trees", Test_tree.tests);
+    ("hashes", Test_hashes.tests);
```
Comment (CraigFe) [inline test/irmin-pack/test_pack.ml:685]: Can we put these first? If these tests fail, I think others are likely to fail as a consequence.

File: test/irmin-pack/common.ml [hunk 8]
```diff
@@ -269,3 +269,19 @@ module Filename = struct
...
+  let counter = ref 0 in
+  iter v (fun x ->
+      match List.nth_opt checks !counter with
+      | None -> Alcotest.failf "No more calls to %s left" iter_type
+      | Some (msg, expected) ->
+          let msg = Fmt.str "Check %s:%s" iter_type msg in
+          check_string ~msg ~expected ~got:x;
+          incr counter);
+  if !counter <> List.length checks then
+    Alcotest.failf "More calls to %s expected" iter_type
```
Comment (CraigFe) [inline test/irmin-pack/common.ml:287]: These seem a bit too specific to me to go in `common`. Can we shift them to `test_hashes`?

Comment (icristescu) [reply test/irmin-pack/common.ml]: good point, done

**Review by Ngoguey42 — APPROVED** [review]
Great, thanks

**Comment by icristescu** [comment]
Thanks for the reviews, merging.

**Pull request merged**